<a href="https://colab.research.google.com/github/VKarpick/powerplay/blob/main/potential_data_issues.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from ast import literal_eval
import pandas as pd

In [2]:
def literal_converter(x):
    return pd.NA if x == "" else literal_eval(x)
    
df = pd.read_csv(
    "https://github.com/VKarpick/powerplay/blob/main/data/powerplay_coords.csv?raw=true",
    converters={
        c: literal_converter
        for c in ("possession_coord", "opponent_coords", "goalie_coord")
    }
)

# Fewer Than 4 PKers in Frame

In [3]:
df["n_opponents"] = df.opponent_coords.apply(len)

In [4]:
missing_players = df.loc[df.n_opponents < 4]
missing_players.set_index("pp_name")
first_frames = missing_players.frame_id - missing_players.frame_id.shift(1) == 1
last_frames = missing_players.frame_id - missing_players.frame_id.shift(-1) == -1

pd.DataFrame({
    "pp_name": missing_players.loc[~first_frames, "pp_name"].values,
    "first": missing_players.loc[~first_frames, "frame_id"].values,
    "last": missing_players.loc[~last_frames, "frame_id"].values,
})

pp_name  first  last
0            2022-02-08 Canada at USA P1 PP2    165   256
1            2022-02-08 Canada at USA P2 PP3   1499  1526
2            2022-02-08 Canada at USA P2 PP3   1541  1585
3            2022-02-08 Canada at USA P2 PP3   2865  2880
4            2022-02-08 Canada at USA P2 PP3   2895  2928
5            2022-02-08 Canada at USA P2 PP5   3172  3175
6           2022-02-08 ROC at Finland P1 PP1    455   570
7           2022-02-08 ROC at Finland P2 PP5   2249  2278
8           2022-02-08 ROC at Finland P3 PP6    872   892
9       2022-02-12 Switzerland at ROC P1 PP1    477   477
10      2022-02-12 Switzerland at ROC P1 PP1   1113  1144
11      2022-02-12 Switzerland at ROC P1 PP2    559   628
12      2022-02-12 Switzerland at ROC P1 PP2   1123  1128
13      2022-02-12 Switzerland at ROC P1 PP2   2345  2388
14      2022-02-12 Switzerland at ROC P3 PP3   3113  3156
15      2022-02-12 Switzerland at ROC P3 PP3   3170  3173
16   2022-02-14 Switzerland at Canada P2 PP4     81   122
17          2022-02-14 USA at Finland P2 PP1    679   723
18          2022-02-14 USA at Finland P2 PP1   2042  2088
19  2022-02-16 Switzerland at Finland P1 PP2   1075  1119
20  2022-02-16 Switzerland at Finland P1 PP2   2589  2603
21  2022-02-16 Switzerland at Finland P2 PP3   1596  1673
22  2022-02-16 Switzerland at Finland P2 PP4   4829  4847
23  2022-02-16 Switzerland at Finland P2 PP5    652   667
24  2022-02-16 Switzerland at Finland P2 PP5    687   707
25  2022-02-16 Switzerland at Finland P3 PP6    740   740
26  2022-02-16 Switzerland at Finland P3 PP7    286   317
27  2022-02-16 Switzerland at Finland P3 PP7    339   387
28  2022-02-16 Switzerland at Finland P3 PP7    398   399
29  2022-02-16 Switzerland at Finland P3 PP8   2301  2359
30  2022-02-16 Switzerland at Finland P3 PP8   2446  2478

Missing backcheckers could affect:
- CAN USA P2 PP3
- SUI ROC P1 PP2
- SUI ROC P3 PP3
- SUI FIN P2 PP3 (maybe)

# Puck Recovery Immediately Follow Possession

In [5]:
possession = pd.read_csv("https://github.com/VKarpick/powerplay/blob/main/data/powerplay_possession.csv?raw=true")

In [6]:
pbp = pd.read_csv("https://github.com/VKarpick/powerplay/blob/main/data/powerplay_pbp.csv?raw=true")

In [7]:
in_zone_possession = possession.copy().loc[possession.is_pp & possession.is_pp_ozone]

recoveries = pbp.copy().loc[pbp.event == "Puck Recovery"]
recoveries["previous_frame_id"] = recoveries.frame_id - 1
recoveries["pp_frame"] = tuple(zip(recoveries.pp_name, recoveries.previous_frame_id))

has_possession = recoveries.pp_frame.isin(zip(in_zone_possession.pp_name, in_zone_possession.frame_id))

recoveries.loc[has_possession, ["pp_name", "frame_id", "clock_seconds", "pp_frame"]]

pp_name  frame_id  clock_seconds  \
15             2022-02-08 Canada at USA P1 PP1     734.0            362   
56             2022-02-08 Canada at USA P1 PP2    1180.0            177   
94             2022-02-08 Canada at USA P1 PP2    2920.0            119   
122            2022-02-08 Canada at USA P2 PP3     467.0            975   
148            2022-02-08 Canada at USA P2 PP3    1783.0            932   
212            2022-02-08 Canada at USA P2 PP5    1734.0            240   
267            2022-02-08 Canada at USA P3 PP6     246.0            536   
424           2022-02-08 ROC at Finland P2 PP5    2661.0            369   
425           2022-02-08 ROC at Finland P2 PP5    2757.0            367   
458           2022-02-08 ROC at Finland P3 PP6     405.0            867   
469           2022-02-08 ROC at Finland P3 PP6    1053.0            845   
558       2022-02-12 Switzerland at ROC P1 PP1    1145.0            710   
590       2022-02-12 Switzerland at ROC P1 PP1    2991.0            649   
616       2022-02-12 Switzerland at ROC P1 PP2     629.0            472   
721       2022-02-12 Switzerland at ROC P3 PP3    2452.0            686   
783       2022-02-12 Switzerland at ROC P3 PP5     503.0             91   
882    2022-02-14 Switzerland at Canada P1 PP1    5627.0            223   
1046   2022-02-14 Switzerland at Canada P3 PP5     249.0            669   
1102   2022-02-14 Switzerland at Canada P3 PP5    3578.0            558   
1112          2022-02-14 USA at Finland P2 PP1     602.0           1048   
1115          2022-02-14 USA at Finland P2 PP1     724.0           1042   
1173          2022-02-14 USA at Finland P2 PP3    1099.0            779   
1180          2022-02-14 USA at Finland P2 PP3    1455.0            767   
1451  2022-02-16 Switzerland at Finland P2 PP3    5180.0           1177   
1468  2022-02-16 Switzerland at Finland P2 PP4     253.0            587   
1708  2022-02-16 Switzerland at Finland P3 PP7     202.0            372   
1792  2022-02-16 Switzerland at Finland P3 PP8    3193.0             86   
1793  2022-02-16 Switzerland at Finland P3 PP8    3266.0             84   
1802  2022-02-16 Switzerland at Finland P3 PP8    3989.0             60   

                                               pp_frame  
15             (2022-02-08 Canada at USA P1 PP1, 733.0)  
56            (2022-02-08 Canada at USA P1 PP2, 1179.0)  
94            (2022-02-08 Canada at USA P1 PP2, 2919.0)  
122            (2022-02-08 Canada at USA P2 PP3, 466.0)  
148           (2022-02-08 Canada at USA P2 PP3, 1782.0)  
212           (2022-02-08 Canada at USA P2 PP5, 1733.0)  
267            (2022-02-08 Canada at USA P3 PP6, 245.0)  
424          (2022-02-08 ROC at Finland P2 PP5, 2660.0)  
425          (2022-02-08 ROC at Finland P2 PP5, 2756.0)  
458           (2022-02-08 ROC at Finland P3 PP6, 404.0)  
469          (2022-02-08 ROC at Finland P3 PP6, 1052.0)  
558      (2022-02-12 Switzerland at ROC P1 PP1, 1144.0)  
590      (2022-02-12 Switzerland at ROC P1 PP1, 2990.0)  
616       (2022-02-12 Switzerland at ROC P1 PP2, 628.0)  
721      (2022-02-12 Switzerland at ROC P3 PP3, 2451.0)  
783       (2022-02-12 Switzerland at ROC P3 PP5, 502.0)  
882   (2022-02-14 Switzerland at Canada P1 PP1, 5626.0)  
1046   (2022-02-14 Switzerland at Canada P3 PP5, 248.0)  
1102  (2022-02-14 Switzerland at Canada P3 PP5, 3577.0)  
1112          (2022-02-14 USA at Finland P2 PP1, 601.0)  
1115          (2022-02-14 USA at Finland P2 PP1, 723.0)  
1173         (2022-02-14 USA at Finland P2 PP3, 1098.0)  
1180         (2022-02-14 USA at Finland P2 PP3, 1454.0)  
1451  (2022-02-16 Switzerland at Finland P2 PP3, 517...  
1468  (2022-02-16 Switzerland at Finland P2 PP4, 252.0)  
1708  (2022-02-16 Switzerland at Finland P3 PP7, 201.0)  
1792  (2022-02-16 Switzerland at Finland P3 PP8, 319...  
1793  (2022-02-16 Switzerland at Finland P3 PP8, 326...  
1802  (2022-02-16 Switzerland at Finland P3 PP8, 398...

Remove possession:
- CAN USA P1 PP1
 - 719 to 733
- CAN USA P1 PP2
 - 1114 to 1179
 - 2853 to 2920
- CAN USA P2 PP3
 - 438 to 466
 - 1743 to 1782
- CAN USA P2 PP5
 - 1689 to 1733
- ROC FIN P2 PP5
 - 2600 to 2660
 - 2662 to 2756
- ROC FIN P3 PP6
 - 375 to 403
 - 976 to 1052
- SUI ROC P1 PP1
 - 1106 to 1144
- SUI ROC P1 PP2
 - 547 to 628
- SUI ROC P3 PP3
 - 2432 to 2451
- SUI ROC P3 PP5
 - 463 to 502
- SUI CAN P1 PP1
 - 5576 to 5626
- SUI CAN P3 PP5
 - 211 to 248
- USA FIN P2 PP1
 - 562 to 601
 - 706 to 723
- USA FIN P2 PP3
 - 1042 to 1098
 - 1238 to 1454
- SUI FIN P2 PP3
 - 5125 to 5179
- SUI FIN P2 PP4
 - 195 to 252
- SUI FIN P3 PP7
 - 162 to 201
- SUI FIN P3 PP8
 - 3107 to 3192
 - 3194 to 3265
 - 3853 to 3988